### PR TITLE
feature: add Curve to Eco Router

### PR DESCRIPTION
**Pull request by adamazad**
# Summary

- Puts Eco Router library logic into a separate contained module. The library manages its own state - it could need a cache layer
- Uses SDK v1.2 Beta 0 with support for native assets on Gnosis Chain trades on Curve. **The contract is pending audit**

**Also in this PR**

- Upgraded styled-components to v5.3.5 and hence you will see many files using the `useTheme` hook. 

If you have reviewed #616 and #781  previously, this should be similar.

### CODE REVIEW

File: src/lib/eco-router/api.ts [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+        errors.push(error)
+        resolve(undefined)
+      })
+  })
+
+  // Wait for all promises to resolve, and
+  // remove undefined values
+  const unsortedTrades = (await Promise.all([...(uniswapV2TradesList as any), curveTrade]).then(trade =>
+    trade.filter(trade => trade !== undefined)
+  )) as Trade[]
```
Comment (wixzi) [inline src/lib/eco-router/api.ts:110]:
```suggestion
  const unsortedTrades: Trade[] = await Promise.all([...(uniswapV2TradesList as any[]), curveTrade]).then(trades =>
    trades.filter(trade => trade !== undefined)
  )
```

File: src/lib/eco-router/api.ts [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+
+import { getUniswapV2PlatformList } from './platforms'
+// Types
+// eslint-disable-next-line
+import type {
+  EcoRouterBestExactInParams,
+  EcoRouterResults,
+  EcoRouterSourceOptionsParams,
+  EcoRouterBestExactOutParams,
+} from './types'
```
Comment (wixzi) [inline src/lib/eco-router/api.ts:15]:
```suggestion
import {
  EcoRouterBestExactInParams,
  EcoRouterResults,
  EcoRouterSourceOptionsParams,
  EcoRouterBestExactOutParams,
} from './types'
```

File: src/lib/eco-router/api.ts [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+      return 0
+    } else {
+      return -1
+    }
+  })
+}
+
+/**
+ * Low-level function to fetch from Eco Router sources
+ * @returns {EcoRouterResults} List of unsorted trade sources
```
Comment (wixzi) [inline src/lib/eco-router/api.ts:43]:
```suggestion
 * @returns {Promise<EcoRouterResults>} List of unsorted trade sources
```

File: src/lib/eco-router/api.ts [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+  // Error list
+  const errors: any[] = []
+  // Derive the chainId from the token in or out
+  const chainId = (currencyAmountIn.currency as Token).chainId ?? (currencyOut as Token).chainId
+
+  // Uniswap V2
+  // Get the list of Uniswap V2 platform that support current chain
+  const uniswapV2PlatformList = getUniswapV2PlatformList(chainId as ChainId)
+
+  const uniswapV2TradesList = uniswapV2PlatformList.map(async platform => {
```
Comment (wixzi) [inline src/lib/eco-router/api.ts:59]: Maybe use a forEach and only push to `uniswapV2TradesList` if undefined

File: src/lib/eco-router/api.ts [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+  // Return the list of sorted trades
+  return {
+    errors,
+    trades: sortTradesByExecutionPrice(unsortedTrades),
+  }
+}
+
+/**
+ * Low-level function to fetch from Eco Router sources
+ * @returns {EcoRouterResults} List of unsorted trade sources
```
Comment (wixzi) [inline src/lib/eco-router/api.ts:121]:
```suggestion
 * @returns {Promise<EcoRouterResults>} List of unsorted trade sources
```

File: src/lib/eco-router/platforms.ts [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+ */
+export function getUniswapV2PlatformList(chainId: ChainId): UniswapV2RoutablePlatform[] {
+  return [
+    UniswapV2RoutablePlatform.SWAPR,
+    UniswapV2RoutablePlatform.UNISWAP,
+    UniswapV2RoutablePlatform.SUSHISWAP,
+    UniswapV2RoutablePlatform.HONEYSWAP,
+    UniswapV2RoutablePlatform.BAOSWAP,
+    UniswapV2RoutablePlatform.LEVINSWAP,
+  ].filter(platform => platform.supportsChain(chainId as ChainId))
```
Comment (wixzi) [inline src/lib/eco-router/platforms.ts:14]:
```suggestion
  ].filter(platform => platform.supportsChain(chainId))
```

**Comment by wixzi** [comment]
Those are optional comments. Feel free to ignore if you feel irrelevent.

**Comment by zamli** [comment]
How about a loading state like this
https://www.figma.com/proto/L8tamoiv3mTCRaypHLk1us/DXswap?page-id=0%3A1&node-id=8677%3A38938&viewport=1663%2C6120%2C1&scaling=min-zoom&starting-point-node-id=8677%3A38938

Design:
https://www.figma.com/file/L8tamoiv3mTCRaypHLk1us/DXswap?node-id=8677%3A38938

**Comment by adamazad** [comment]
The hooks consume a lot of resources by searching many liquidity sources. React hooks should abort running the Eco Router when:

1. User wants to wrap the native network asset and vice versa
2. User chosen input currency balance is sufficient to conclude a trade

**Comment by adamazad** [comment]
@MilanVojnovic95 can you try to break this? :D

File: src/lib/eco-router/api.ts (hunk 5 above)
Comment (adamazad) [reply src/lib/eco-router/api.ts]: Addressed in [da93579](https://github.com/levelkdev/swapr-dapp/pull/977/commits/da93579f04388949d8a594a4c468853147648947)

File: src/components/Header/index.tsx [hunk 7]
```diff
@@ -241,17 +241,24 @@ function Header() {
...
       }
-    })
+    }
+  }
+
+  useEffect(() => {
+    window.addEventListener('scroll', onScrollHander)
+
+    return () => {
+      console.log('unmounting')
```
Comment (berteotti) [inline src/components/Header/index.tsx:259]: WE don't need this console.log :)

```suggestion
```

File: src/components/SelectionList/SelectionList.tsx [hunk 8]
```diff
@@ -1,18 +1,19 @@
 import { Text } from 'rebass'
-import styled from 'styled-components'
+import styled from 'styled-components/macro'
```
Comment (berteotti) [inline src/components/SelectionList/SelectionList.tsx:2]: Should we use `'styled-components/macro'`import everywhere? Or should we use `'styled-components'` here?

File: src/components/swap/SwapPlatformSelector.tsx [hunk 9]
```diff
@@ -155,54 +159,58 @@ export function SwapPlatformSelector({
...
+                    {tokenAmount === '0' ? '<0.0000001' : tokenAmount || '-'}
+                  </TYPE.subHeader>
+                  <CurrencyLogo
+                    currency={isExactIn ? trade.outputAmount.currency : trade.inputAmount.currency}
+                    size="14px"
+                    marginLeft={4}
+                  />
+                </SelectionListReceiveAmount>
+              </SelectionListOption>
+            )
```
Comment (berteotti) [inline src/components/swap/SwapPlatformSelector.tsx:211]: As I'm looking into this code, I'm thinking about turning this into a new component to improve readability. It is just a suggestion, we could make an issue and solve it latter if you think it should belong in another PR/task.

File: src/components/SelectionList/SelectionList.tsx (hunk 8 above)
Comment (adamazad) [reply src/components/SelectionList/SelectionList.tsx:2]: I used to debug a UI glitch. Switching to macro is on the [task list for July](https://www.notion.so/adamazad/Swapr-UX-DX-Improvement-Plans-for-July-2022-v1-0-904ba420c12e4014831ea9e1ed05ae05). For now, I don't think we should make this PR any bigger in terms of the file list.

File: src/components/swap/SwapPlatformSelector.tsx (hunk 9 above)
Comment (adamazad) [reply src/components/swap/SwapPlatformSelector.tsx:211]: I have the same concern about readability here. I created a ticket: #1015, and we can do that next.

File: src/hooks/useSwapCallback.ts [hunk 10]
```diff
@@ -99,13 +99,49 @@ export function useSwapsCallArguments(
...
+    platformName !== UniswapV2RoutablePlatform.SWAPR.name ? `on ${platformName}` : ''
+  }`
+
+  return recipientAddressOrName != null
+    ? `${base} to ${
+        recipientAddressOrName && isAddress(recipientAddressOrName)
+          ? shortenAddress(recipientAddressOrName)
+          : recipientAddressOrName
+      }`
+    : base
```
Comment (berteotti) [inline src/hooks/useSwapCallback.ts:122]: Can we add translations here?

File: src/pages/Swap/index.tsx [hunk 11]
```diff
@@ -406,85 +528,7 @@ export default function Swap() {
...
-                      }}
-                      id="swap-button"
-                      disabled={!isValid || (priceImpactSeverity > 3 && !isExpertMode) || !!swapCallbackError}
-                      platformName={trade?.platform.name}
-                      swapInputError={swapInputError}
-                      priceImpactSeverity={priceImpactSeverity}
-                      isExpertMode={isExpertMode}
-                    ></SwapButton>
-                  )}
+                  <SwapBoxButton />
```
Comment (berteotti) [inline src/pages/Swap/index.tsx:531]: This is definitely better!

File: src/lib/eco-router/hooks.ts [hunk 12]
```diff
@@ -0,0 +1,165 @@
...
+
+    return () => {
+      isCancelled = true
+    }
+    // eslint-disable-next-line
+  }, [
+    chainId,
+    uniswapV2IsMultihop,
+    // eslint-disable-next-line
+    currencyAmountIn?.toSignificant(),
```
Comment (berteotti) [inline src/lib/eco-router/hooks.ts:78]: Why are we doing this? Should we use a `useRef` in this case?

File: src/pages/Swap/index.tsx [hunk 13]
```diff
@@ -297,6 +311,114 @@ export default function Swap() {
...
+      )
+    }
+
+    if (showApproveFlow) {
+      return (
+        <RowBetween>
+          <ButtonConfirmed
+            onClick={approveCallback}
+            disabled={approval !== ApprovalState.NOT_APPROVED || approvalSubmitted}
+            width="48%"
```
Comment (jorgedclopes) [inline src/pages/Swap/index.tsx:354]: Just a reminder... We wanted to let the Buttons address these properties themselves instead of hardcoding them.
Non-blocking.

File: src/lib/eco-router/hooks.ts (hunk 12 above)
Comment (adamazad) [reply src/lib/eco-router/hooks.ts:78]: `currencyAmountIn` is a new object every time, causing infinite re-renders. Comparing the sting value is prevents this.

File: src/hooks/useSwapCallback.ts [hunk 14]
```diff
@@ -99,13 +99,49 @@ export function useSwapsCallArguments(
...
+  const outputSymbol = trade.outputAmount.currency.symbol
+  const inputAmount = trade.inputAmount.toSignificant(3)
+  const outputAmount = trade.outputAmount.toSignificant(3)
+  const platformName = trade.platform.name
+
+  const base = `Swap ${inputAmount} ${inputSymbol} for ${outputAmount} ${outputSymbol} ${
+    platformName !== UniswapV2RoutablePlatform.SWAPR.name ? `on ${platformName}` : ''
+  }`
+
+  return recipientAddressOrName != null
```
[the remark this answers is not in the document]
Comment (adamazad) [reply src/hooks/useSwapCallback.ts:116]: That's the mistake, should be `string | null`

File: src/hooks/useSwapCallback.ts [hunk 15]
```diff
@@ -133,6 +169,8 @@ export function useSwapCallback(
     return {
       state: SwapCallbackState.VALID,
       callback: async function onSwap(): Promise<string> {
+        // eslint-disable-next-line
+        // @ts-ignore
```
[the remark this answers is not in the document]
Comment (adamazad) [reply src/hooks/useSwapCallback.ts:173]: I believe this is a residue from my previous work. Thanks!

File: src/lib/eco-router/api.ts [hunk 16]
```diff
@@ -0,0 +1,208 @@
...
+        errors.push(error)
+        resolve(undefined)
+      })
+  })
+
+  // Wait for all promises to resolve, and
+  // remove undefined values
+  const unsortedTrades: Trade[] = await Promise.all([...(uniswapV2TradesList as any[]), curveTrade]).then(trades =>
+    trades.filter(trade => trade !== undefined)
+  )
```
[the remark this answers is not in the document]
Comment (adamazad) [reply src/lib/eco-router/api.ts]: Addressed in [d8eb364](https://github.com/levelkdev/swapr-dapp/pull/977/commits/d8eb3648d409491a46a0ee26dd8d5922fb4a10c7)

File: src/lib/eco-router/api.ts [hunk 17]
```diff
@@ -0,0 +1,208 @@
...
+
+  // Return the list of sorted trades
+  return {
+    errors,
+    trades: sortTradesByExecutionPrice(unsortedTrades),
```
[the remark this answers is not in the document]
Comment (adamazad) [reply src/lib/eco-router/api.ts]: Addressed in [d8eb364](https://github.com/levelkdev/swapr-dapp/pull/977/commits/d8eb3648d409491a46a0ee26dd8d5922fb4a10c7)

File: src/hooks/useSwapCallback.ts [hunk 18]
```diff
@@ -99,13 +99,49 @@ export function useSwapsCallArguments(
...
+    ? `${base} to ${
+        recipientAddressOrName && isAddress(recipientAddressOrName)
+          ? shortenAddress(recipientAddressOrName)
+          : recipientAddressOrName
+      }`
+    : base
+}
+
+export interface UseSwapCallbackParams {
+  trade: Trade | undefined // trade to execute, required
```
[the remark this answers is not in the document]
Comment (adamazad) [reply src/hooks/useSwapCallback.ts]: Addressed in [d8eb364](https://github.com/levelkdev/swapr-dapp/pull/977/commits/d8eb3648d409491a46a0ee26dd8d5922fb4a10c7)

**Comment by MilanVojnovic95** [comment]
Retested and looks ok to me 👍

**Review by denciu — APPROVED** [review]
